### ISSUE TITLE
Patients in a loop of loo use, keep on re-entering the Toilet room

### ISSUE THREAD

**Issue by TheCycoONE** [CorsixTH/CorsixTH#555]
The issue is still relevant for 2026.

----

### Describe the issue

Patients in a loop of loo use, keep on re-entering the Toilet room

----

https://github.com/user-attachments/assets/ad0823d5-cc79-4ade-825f-510b399a3500

----

### Save Game

[toilet-test.zip](https://github.com/user-attachments/files/25804645/toilet-test.zip)

----

### CorsixTH Version

trunk master 0.70.0 53d8091

----

### Steps to Reproduce

1. Load attached save game. 
2. Observe that there is a new toilet in the upper, northwest part of the hospital. Its queue is empty yet. Several patients are heading to it.
3. Observe that when these patients start to use that room, after use and exit the room patients will go back to the toilet queue. So they get stuck here forever, creating a never-ending queue.

----

### Expected Behaviour

After using the toilet, patients should go on with their business, move further into diagnosis and treatment, rather than queueing for the same toilet again.

----

Earlier issue description (2014):

_From [wolfy1...@gmail.com](https://code.google.com/u/111955299291263404988/) on May 07, 2013 17:33:54_

What steps will reproduce the problem? 1.i am not sure how to reproduce this problem. 2. 3. What is the expected output? What do you see instead? A patient that just used the toilets returns even though he does not need to.

What version of CorsixTH are you using (e.g. "Beta 2", "SVN r123 ")?
0.21 What operating system / compile settings are you using? Windows 8 Pro 64 bit
What level was this on (e.g. "Demo level", "Full game level 12")?
Full Game level Attach a saved game file (preferably an autosave from before the error) if you have one and it is relevant. Please provide any additional information

**Attachment:** [config.txt gamelog.txt thomas01.sav](http://code.google.com/p/corsix-th/issues/detail?id=1623)

_Original issue: http://code.google.com/p/corsix-th/issues/detail?id=1623_

**Comment by Glimmer-Man** [comment]
My clue: this will happen when there are more patients trying to get on toilet then number of toilets in the room.

Next clue: ...maybe it's connected with pateints which will pee outside while they are in toilet queue.

I've made may own game change, so I'm playing with random number of spawn patients (1, 1500). So there are moments, when I have 1000 - 1300 patients in hospital, and moments when almost nobody is there.

And this issue happens only when big crowd is everyvere...

Version: 0.40

**Comment by MarkL1961** [comment]
It should be free toilets that limits who can enter the room.  Peeing outside is a possibility i.e. on the way and don't quite make it..
Agree with what you say about amount of patients overall, the only times I  have seen this problem is with overcrowded hospital, otherwise I don't see this problem.  You also get lots of other errors though when the hospital is overcrowded.

**Comment by wolfy1339** [comment]
I haven't been able to reproduce this myself lately.

**Comment by MarkL1961** [comment]
Perhaps sdl2 is more stable!
Have you had large numbers of patients, as that seemed to be related?

**Comment by wolfy1339** [comment]
Haven't tested on SDL2 (I haven't been able to compile) only 0.40 but on a different save.
Does anyone have a copy of the save or know how to get it?

**Comment by MarkL1961** [comment]
The original link above was meant to work, but @Lego3  ticked something on googlecode that now means you cannot access it.  I noticed the same applies to other links we have in the suggestions wiki etc.

I don't think this was intended, so I wonder if it is worth removing that for the time being Edvin?

**Comment by Lego3** [comment]
Ok, let's have it open for a while longer. I would just like to close the google code project to avoid people reading outdated things there.

**Comment by wolfy1339** [comment]
I have downloaded all the attachments onto my server,
thomas01.sav: http://vps3.wolfy1339.tk/~wolfy1339/CorsixTH/issue555/thomas01.sav
gamelog.txt: http://vps3.wolfy1339.tk/~wolfy1339/CorsixTH/issue555/gamelog.txt
config.txt: http://vps3.wolfy1339.tk/~wolfy1339/CorsixTH/issue555/config.txt

I can do the same for the other issues too, if needed.

**Comment by Glimmer-Man** [comment]
New final (I hope) clue:

Example: 5 patients arrive to toilet and they makes queue outside of door. If the first one pee on floor, it's OK, nothing happen. But if second one (or other after) pee on the floor, it makes him try to find another toilet room. And that's the source of problem because he will spend rest of his life finding toilet room (if I will not kick him out from hospital).

Fact 1) 
It's happenig only when there is a long queue.

Fact 2)
Patient with toilet paper above his head when he's entering toilet room will not do that. So It must be connected with peeing on the floor.

I hope that my clue it The One. :-)

**Comment by mugmuggy** [comment]
It appears to be, that Room:tryToFindNearbyPatients is called for the toilet. That long queue is a good indicator to when tryToFindNearbyPatients will run and perform the tryMovePatient code.

This code will automatically set next_room_to_visit if it does calculate that a move should occur, and upon doing so it will put it in a loop via the ToiletRoom:dealtWithPatient code. Either could avoid tryToFindNearbyPatients (by overriding method), or have dealtWithPatient swing them by reception again, and get them to gps office to start again. if they are trying to find the toilets (also would need reception not to send them to the toilets).

**Comment by mugmuggy** [comment]
Just to clarify the situation, the tryMovePatient guards against the redirection occuring if patient.going_to_toilet == "yes" essentially. However if the queues are long enough, the patient going to the toilet whilst in the queue (fact 2) will reset them to "no" and they will fall into this redirection loop. I can't recall what the original did, if the patient ended up having the toilet need go, whilst they were on the way to the toilet, if they remained continuing on, or if they proceeded to the next step.

Also, if the patient is seated, and they get the call to Patient:pee this incorrectly executes and resets the no value, and sets the toilet_need to a lower amount than the threshold. So currently benches reduce toilet need.

**Comment by mugmuggy** [comment]
Just also noting with this code
https://github.com/CorsixTH/CorsixTH/blob/1dd16721273d65ac81053748d8ef7ef7eccc196a/CorsixTH/Lua/entities/humanoids/patient.lua#L791-L794

The patients toilet need is adjusted prior to the actions completing, whereas when they use the toilets its deferred to the room to handle it after toilet use.

**Comment by lewri** [comment]
May be fixed by PR #1565

**Comment by lewri** [comment]
I can't seem to replicate this in 0.64...

mugmuggy did you have any input on if this still happens?

**Comment by mugmuggy** [comment]
None of the changes I made corrected this. I did look at trying to get the going_to_toilet flag to persist nicely but it did introduce a side effect but I can't recall exactly what that was but it was separate to the PR at the time and might have just required more thought.

To replicate need to have many patients wanting to use the toilet. They need to queue at the toilet and be the 2nd or later in queue position (to enter). Whilst in the queue they need to pee, before entering the toilet.

At the same time you want to pause (just to get timing right) and build another toilet, which will do the Room:tryToFindNearbyPatients call which sets the next_room_to_visit for the patient in tryMovePatient, which the guard for going to toilet is now bypassed due to the pee call occuring.

I replicated this in the save from 1455. You'll see patients with toilet need icons, who have a toilet need less than threshold that are seated. Because they are seated the pee action doesn't complete, build a toilet and a handful or so stream in.

But make Patient:pee return true only if the actions were queued, and check for that in patient tickDay - the code in my previous comment, don't adjust need/going_to_toilet - build toilet with a single cubicle and you'll have 40+ patients queue. Some of these patients will pee now in the queue, and thus give you a chance to build another toilet, that they hopefully get redirected and enter the loop. Obviously this is manipulating the state, to support what can happen randomly.

It is more than likely only going to happen when a new room is built (ref 1337) as the getUsageScore only takes into account the currently queued count, and nothing with the expected (which was broken anyway, ref 1568). Likewise the patients probably shouldn't pee whilst walking to a toilet/queuing for a toilet or if they do, skip going to the toilet and resume at the next point.

**Comment by Matroftt** [comment]
i remember i had this situation in one of my saves.. going to send it there if i find it

**Comment by lewri** [comment]
Replicated again now. Save attached.

[toilet-test.zip](https://github.com/user-attachments/files/25804645/toilet-test.zip)

**Comment by TheCycoONE** [comment]
Oh, this is a classic, good work

**Comment by ARGAMX** [comment]
So, here are the first interesting results of this bug investigation.

Seems the mechanism for sending a patient to the toilet works through an "interruption".

For example, a patient is standing in line for a certain room. They are currently busy trying to get to that target room.
A specific mechanism periodically checks whether the patient needs to go to the toilet.
If it's time to go to the toilet, an "interruption" occurs.

At the moment of this "interruption", the current target room is written to a temporary variable, and the new target room becomes the toilet. This "temporary variable" stores only one value.
And the patient is sent on their journey to the toilet.

When the patient finishes their business in the toilet, a mechanism is triggered to return to the state before they were sent to the toilet—the previous target room from "temporary variable" is set as target room.
And the patient goes back to the room they were trying to get to when the "interruption" occurred.

https://github.com/CorsixTH/CorsixTH/blob/1bbafd50296c57af475789504352506b7e5e76a4/CorsixTH/Lua/rooms/toilets.lua#L61-L71

This is how the logic of visiting the toilet works for patients in general.
I described it with some assumptions, but in general the essence is this.
Can you already sense the weak point here?

[spoiler]
The weak point is that if the interrupt mechanism triggers again while the patient's target room is the toilet, the "temporary variable" storing the room to return to after using the toilet will be overwritten with the toilet.

This way, the patient will have the following state:
now going - to the toilet.
after using the toilet, return - to the toilet.
And thus, the patient gets caught in an endless loop of living in the toilet. 🙂

It seems that if this happens, the patient will never be able to return to normal life on their own. At least, I haven't yet been able to break such patients out of this loop.

Even if you remove all the toilets... for such patients the stack overflow will happen and Lua will crash. Hilarious.

```Yaml
../Resources/Lua/world.lua:1676: stack overflow
stack traceback:
	../Resources/Lua/world.lua:1676: in method 'findRoomNear'		
	../Resources/Lua/humanoid_actions/seek_toilets.lua:43: in field '?'	
	../Resources/Lua/entities/humanoid.lua:596: in method 'startAction'
	../Resources/Lua/entities/humanoid.lua:702: in method 'finishAction'
	../Resources/Lua/humanoid_actions/seek_room.lua:206: in field '?'
	../Resources/Lua/entities/humanoid.lua:596: in method 'startAction'
	../Resources/Lua/entities/humanoid.lua:702: in method 'finishAction'
	../Resources/Lua/humanoid_actions/seek_toilets.lua:67: in field '?'
	../Resources/Lua/entities/humanoid.lua:596: in method 'startAction'
	../Resources/Lua/entities/humanoid.lua:702: in method 'finishAction'
	...	(skipping 315753 levels)
	../Resources/Lua/entities/humanoid.lua:596: in method 'startAction'	
	../Resources/Lua/entities/humanoid.lua:702: in method 'finishAction'	
	../Resources/Lua/humanoid_actions/seek_room.lua:206: in field '?'	
	../Resources/Lua/entities/humanoid.lua:596: in method 'startAction'	
	../Resources/Lua/entities/humanoid.lua:702: in method 'finishAction'	
	../Resources/Lua/humanoid_actions/walk.lua:192: in local 'timer_function'
	../Resources/Lua/entity.lua:222: in field 'tick'
	../Resources/Lua/entities/humanoids/patient.lua:599: in method 'tick'
	../Resources/Lua/world.lua:915: in method 'onTick'
	../Resources/Lua/app.lua:1305: in function <../Resources/Lua/app.lua:1302>
	(...tail calls...)
	../Resources/Lua/app.lua:1205: in function <../Resources/Lua/app.lua:1200>
```

I also haven't yet figured out how to make the interrupt trigger a second time while the patient is heading to the toilet. So, I can't yet reproduce the situation from scratch. But I'm working on it.

Interrupt trigger a second time is definitely something that the developer of this mechanism did not plan, but nevertheless it happens and happened in the Lewri save game.

...
Oh, right, I think mugmuggy described how to replicate this above—you need to have a patient urinate on the floor while standing in line for the toilet. And then build another toilet. I'll check it out.

**Comment by ARGAMX** [comment]
Perhaps, when exiting the toilet, when reading `next_room_to_visit`, we should check that it's not a toilet.
If it is a toilet, then queueAction `SeekReceptionAction()` or some other safe fallback.
This way, the consequence, not the cause, will be eliminated.

**Comment by ARGAMX** [comment]
Why this defect happens?
In the cycle of this defect, `toilet` ends up in `next_room_to_visit`.
When the patient leaves the `toilet`, they checks `next_room_to_visit`to back to main flow.
But there is a `toilet` in `next_room_to_visit`, so they goes to the `toilet` again.
And so on, ad infinitum. 

Generally, it’s obvious that only diagnostic and treatment rooms should be stored in `next_room_to_visit`.
That is, `next_room_to_visit` should help the patient return to the main scenario of the patient (to be cured), not to some extraneous goals (to pee in toilet).
By interruptions, I mean: using the drink vending machine, throwing up, urinating on the floor, and so on.

So today it occurred to me that we just need to prohibit putting `toilet` in `next_room_to_visit`. 
Then the defect chain will break.
This won’t help fix the already defective saves. 
It will help avoid falling into a defective cycle.

However, it might break something.
I don’t know exactly what.
But why something put `toilet` in `next_room_to_visit`?
I think some kind of interruption is doing this.
And if we stop putting `toilet` in `next_room_to_visit`, then after such an interruption, possible, the patient won’t be able to go further to the toilet and will have to back to treatment again.
In short, the patient after using vending machine may forget where they were going and forget to get to the toilet. A kind of ADHD person.
That’s about the only consequence I can see.
Well, this is clearly better than getting stuck in an endless loop with the toilet.

Should I open such a PR?

Well, in general, to solve the problem more effectively without consequence, we need to develop a new action management system.
For example, `Humbe`, as Albert suggests.
